### ISSUE TITLE
Building PTB LSTM Language Model in TFLearn

### ISSUE THREAD

**Issue by krayush07** [tflearn/tflearn#209]
How can we model the PTB example in Tensorflow (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/models/rnn/ptb/ptb_word_lm.py) in TFLearn.

As of now, all the LSTM models given in the example in TFLearn uses the prediction only from the final LSTM unit to calculate the error. However, in learning sequence, it is required to compute the error over the whole sequence (other metrics such as perplexity also required logits over the whole sequence rather than just final LSTM prediction).

What is the best way to achieve this i.e, to calculate error and perplexity over the whole sequence.

**Comment by aymericdamien** [comment]
Yes, maybe you need to use a custom loss function like this one (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/models/rnn/ptb/ptb_word_lm.py#L129) , we should check if this is directly compatible with TFLearn regression layer or not (and if not, adapt it to make it works).
If it is not working, you may try to change the loss function here (https://github.com/tflearn/tflearn/blob/master/tflearn/layers/estimator.py#L139).
I will try to work on that too when I get some free time. This will be very useful to make that works! Seq2seq example is kinda missing in TFLearn now.

**Comment by krayush07** [comment]
Yes, I was trying to check but the blocker that I faced was that as soon as regression layers is constructed, it expects a loss function and internally the model is trained for just layer/LSTM unit loss. Could you suggest a flow/method to try it out, so that I can look for it too?

**Comment by aymericdamien** [comment]
I think you may try to implement your own regression layer; I think you just need to use that loss function (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/models/rnn/ptb/ptb_word_lm.py#L129) and prior concatenate your outputs, similar to the TF seq2seq examples. Here is the function description: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/seq2seq.py#L890.
Basically I think you just need to concatenate all outputs from the RNN layer (that you get with return_seq=True), and then feed them into that loss function. This function will apply a softmax_crossentropry at each timestep.
